api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+		input.SetDesiredStatus(status)
+		input.SetStartedBy(startedBy)
+		if err := e.Client.ECS.ListTasksPages(input, fn); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, taskARN := range taskARNs {
+		task, err := e.TaskProvider.Read(taskARN)
+		if err != nil {
+			return nil, err
+		}
+
+		taskResourceConsumers, err := e.getContainerResourceFromDeploy(task.DeployID)
+		if err != nil {
+			return nil, err
+		}
+
+		resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers_TasksInJobs(clusterName string) ([]ResourceConsumer, error) {
+	var resourceConsumers []ResourceConsumer
+	jobs, err := e.JobStore.SelectAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, job := range jobs {
+		if job.Type == models.CreateTaskJob {
+			// TODO: maybe remove pending check here
+			if job.Status == models.PendingJobStatus || job.Status == models.InProgressJobStatus {
+				var req models.CreateTaskRequest
+				if err := json.Unmarshal([]byte(job.Request), &req); err != nil {
+					return nil, err
+				}
+
+				if req.EnvironmentID == clusterName {
+					// note that this isn't exact if the job has started some, but not all of the tasks
+					taskResourceConsumers, err := e.getContainerResourceFromDeploy(req.DeployID)
+					if err != nil {
+						return nil, err
+					}
+
+					resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+				}
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*ResourceProvider, error) {
+	var result []*ResourceProvider
+
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
+	listContainerInstancesInput.SetCluster(clusterName)
+	listContainerInstancesInput.SetStatus(ecs.ContainerInstanceStatusActive)
+
+	containerInstanceARNs := []*string{}
+	listContainerInstancesPagesFN := func(output *ecs.ListContainerInstancesOutput, lastPage bool) bool {
+		containerInstanceARNs = append(containerInstanceARNs, output.ContainerInstanceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListContainerInstancesPages(listContainerInstancesInput, listContainerInstancesPagesFN); err != nil {
+		return nil, err
+	}
+
+	if len(containerInstanceARNs) == 0 {
+		return result, nil
+	}
+
+	describeContainerInstancesInput := &ecs.DescribeContainerInstancesInput{}
+	describeContainerInstancesInput.SetCluster(clusterName)
+	describeContainerInstancesInput.SetContainerInstances(containerInstanceARNs)
+
+	output, err := e.Client.ECS.DescribeContainerInstances(describeContainerInstancesInput)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(output.ContainerInstances) == 0 {
+		return result, nil
+	}
+
+	for _, instance := range output.ContainerInstances {
+		// it's non-intuitive, but the ports being used by the tasks live in
+		// instance.RemainingResources, not instance.RegisteredResources
+		var (
+			usedPorts       []int
```

Comment:
don't use this style of declaration

Reply:
aren't the `defaultPorts` always in use here?

Reply:
re vars: ✅ 

re `defaultPorts`: I think so, but we also might have consumers using ports beyond the defaults as well. Still seems to me to be cleaner to just grab all and `int`ify all the ports. /shrug